fdbserver/storageserver.actor.cpp

```diff
@@ -876,6 +886,295 @@ struct StorageServer {
+						if (trace)
+							TraceEvent("Nim_case 1").detail("rangeReadNum", rangeReadNum);
+						if (kvOpt.present()) {
+							if (trace)
+								TraceEvent("Nim_case 1 has")
+								    .detail("key", kvOpt.get())
+								    .detail("rangeReadNum", rangeReadNum);
+							result.data.push_back_deep(result.arena, kvOpt.get());
+							decrementLimits(kvOpt.get());
+						}
+						return true;
+					}
+
+					if (!kvOpt.present()) {
+						// If kvOpt is empty then there are no more results on disk so we continue iterate through only
+						// the ptree and adding relevant results
+						finishedDisk = true;
+						if (vCurrent->isClearTo()) { // dont add the key if its part of a clear range
+							++vCurrent;
+							continue;
+						}
+						if (trace)
+							TraceEvent("Nim_case 2").detail("key", vCurrent.key()).detail("rangeReadNum", rangeReadNum);
+						KeyValueRef kvRef = KeyValueRef(vCurrent.key(), vCurrent->getValue());
+						result.data.push_back_deep(result.arena, kvRef);
+						decrementLimits(kvRef);
+						++vCurrent;
+						continue;
+					}
+
+					ASSERT(!finishedDisk);
+					KeyValueRef kv = kvOpt.get();
+					if (kv.key < vCurrent.key()) {
+						// If the key from the storage engine is smaller than the ptree key
+						// then add the storage engine key to the results
+						if (trace)
+							TraceEvent("Nim_case 3")
+							    .detail("key", kv.key)
+							    .detail("rangeReadNum", rangeReadNum)
+							    .detail("vCurrent", vCurrent.key())
+							    .detail("clearTo", vCurrent->isClearTo());
+						result.data.push_back_deep(result.arena, kv);
+						decrementLimits(kv);
+						return true;
+					} else if (vCurrent->isClearTo()) {
+						// Encounter a clear range in the ptree
+						KeyRef skipKey = vCurrent->getEndKey(); // get the end of the clear range
+						if (trace)
+							TraceEvent("Nim_case 4 skipKey")
+							    .detail("curKey", vCurrent.key())
+							    .detail("key", skipKey)
+							    .detail("rangeReadNum", rangeReadNum)
+							    .detail("disk", kv.key);
+						++vCurrent;
+						// If the end of the clear range is greater than the key from disk then tell the storage engine
+						// to skip to the end of the clear range (otherwise keep iterating through the ptree)
+						if (!vCurrent || vCurrent.key() > kv.key)
+							return skipKey;
+						if (trace)
+							TraceEvent("Nim_case 4 vCurrent")
+							    .detail("key", vCurrent.key())
+							    .detail("rangeReadNum", rangeReadNum)
+							    .detail("clearTo", vCurrent->isClearTo())
+							    .detail("disk", kv.key);
+					} else if (kv.key == vCurrent.key()) {
+						// If the key on disk equals the key in the ptree then use the one from the ptree (as its more
+						// recent)
+						KeyValueRef kvRef = KeyValueRef(vCurrent.key(), vCurrent->getValue());
+						if (trace)
+							TraceEvent("Nim_case 5").detail("key", kvRef).detail("rangeReadNum", rangeReadNum);
+						result.data.push_back_deep(result.arena, kvRef);
```

Comment:
```suggestion
						result.data.emplace_back_deep(result.arena, KeyValueRef(vCurrent.key(), vCurrent->getValue()));
```
I think you can avoid a copy here by moving the new KVRef and deleting the `kvRef` variable above.